docs/proposals/scheduling/20220609-resource-reservation.md

```diff
@@ -0,0 +1,253 @@
+	LabelSelector *metav1.LabelSelector  `json:"labelSelector,omitempty"`
+}
+
+type ReservationPhase string
+
+const (
+	ReservationPhasePending   ReservationPhase = "Pending"
+	ReservationPhaseAvailable ReservationPhase = "Available"
+	ReservationPhaseWaiting   ReservationPhase = "Waiting"
+	ReservationPhaseExpired   ReservationPhase = "Expired"
+)
+
+type ReservationCondition struct {
+	LastProbeTime      metav1.Time `json:"lastProbeTime"`
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	Reason             string      `json:"reason"`
+	Message            string      `json:"message"`
+}
+```
+
+### Implementation Details
+
+#### Reservation Plugin
+
+##### Schedule Reservations
+
+A `Reservation` object has its scheduling requirements like a pod. Ideally, A `Reservation` object should get processed directly by the scheduler like a pod. However, it can require a series of modifications on [scheduling framework](#glossary), losing the compatibility with standard kube-scheduler, kubelet, autoscaler, etc. In the reservation plugin, we fake one *reservation pod* for one `Reservation` inside the scheduler to fulfill general scheduling plugins (noderesources, nodeaffinity, tainttolerations, ...). The scheduling framework can handle `Reservation` objects by processing fake pods in both [scheduling cycle and binding cycle](https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/#scheduling-cycle-binding-cycle).
+
+A fake pod inside the scheduler can construct the same affinity/anti-affinity constraints as owner pods, which may change the reservation result. To handle this problem, koord-scheduler extends the framework to skip check of pod affinity for existing reservations in the `Filter` phase.
+
+A reservation specified `PreAllocation` intends to pre-allocate resources on nodes. The scheduler will skip its filtering of node resources in the scheduling cycle. However, the scheduled reservation will be `Waiting` to be `Available` until there are enough resources to fulfill its requests.
+
+If all nodes are unscheduled for the reservation, the scheduler keeps its status as `Pending` and sets `Conditions` with the failure message.
+
+Once the scheduling decision has been made, the corresponding `Reservation` object is updated with a new status indicating whether the reservation succeeded or not. The fake pod does not expose to other components, and the kubelet without modification does not perceive a `Reservation` assigned. Fortunately, a `Reservation` does not need to be executable on the node, so existing containers can keep running as usual without additional admissions.
+
+If a reservation has set the `nodeName` (inside the `template` field), the scheduler is responsible for checking if the node can fulfill the reservation since kubelet does not do admissions for the reservation.
+
+##### Allocate Reserved Resources
+
+Let's call the reservation is *allocatable* for a pod iff:
```

Comment:
`iff` -> `if`

Reply:
Done. Anyway, you know what I mean - -.

Reply:
![image](https://user-images.githubusercontent.com/24452340/174217927-e0ef39c7-c561-44be-b949-682b0346e720.png)